src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public sealed class HeaderMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, IEndpointSelectorPolicy
```

Comment:
We probably want to make this an IEndpointNodeBuilder too, but that's more of an advanced topic.

Reply:
shall I file an issue to track? We don't see this as a requirement atm.

Reply:
It will provide more optimized matching. Most of the policies in routing implement this (httpmethod, host, etc). I'm fine if we file an issue and leave it as a follow-up.

Reply:
Agreed. I will file an issue to add INodeBuilderPolicy support, and we can tackle it separately.